crates/ironrdp-pdu/README.md

```diff
@@ -1,3 +1,133 @@
 # IronRDP PDU
 
-RDP PDU encoding and decoding.
+RDP PDU encoding and decoding library.
+
+## WIP: overview of encoding and decoding traits
+
+It’s important for `PduEncode` to be object-safe in order to enable patterns such as the one
+found in `ironrdp-svc`:
+
+```rust
+pub trait StaticVirtualChannel {
+    fn process(&mut self, payload: &[u8]) -> PduResult<Vec<Box<dyn PduEncode>>>;
+}
+```
+
+(The actual trait is a bit more complicated, but this gives the rough idea.)
+
+TODO: elaborate this section
+
+## Difference between `WriteBuf` and `WriteCursor`
+
+`WriteBuf` is used in APIs where the required space cannot be known in advance. For instance,
+`ironrdp_connector::Sequence::step` is taking `&mut WriteBuf` instead of `&mut
+WriteCursor` because it’s unlikely the user knows exactly how much space is required to encode a
+specific response before actually processing the input payload, and as such there is no way
+to communicate the caller how much space is required before calling the function.
+
+`WriteCursor`, in essence, is a helper for this kind of code:
+
+```rust
+    let pdu = /* some PDU structure */;
+
+    // Find the required space
+    let pdu_size = pdu.size();
+
+    // Resize the buffer, making sure there is enough space to fit the serialized PDU
+    if buf.len() < pdu_size {
+        buf.resize(pdu_size, 0);
+    }
+
+    // Proceed to actually serialize the PDU into the buffer…
+    let mut cursor = WriteCursor::new(dst);
+    encode_cursor(pdu, &mut cursor)?;
+```
+
+But it’s going a bit further by keeping track of the already filled region (as opposed to the
+initialized yet unfilled region of the inner `Vec<u8>`). One purpose is to facilitate buffer re-
+use: it’s okay to keep around the same `WriteBuf` and repeatedly write into it. Contrary to
+`Vec<u8>`, bytes will be written at the beginning of the unfilled region (a `Vec<u8>` will write
+bytes at the beginning of the uninitialized region). This way, it’s easy and safe to retrieve a
+mutable slice of the unfilled (but already initialized) region. This slice can be used to build
+a `WriteCursor`. Another purpose of the filled region tracking is to enable multiple items to be
+written consecutively in the same buffer. For instance, the `encode_buf` function looks like this:
+
+```rust
+pub fn encode_buf<T: PduEncode + ?Sized>(pdu: &T, buf: &mut WriteBuf) -> PduResult<usize> {
+    let pdu_size = pdu.size();
+    let dst = buf.unfilled_to(pdu_size);
+    let mut cursor = WriteCursor::new(dst);
+    encode_cursor(pdu, &mut cursor)?;
+    let written = cursor.pos();
+    buf.advance(written);
+    Ok(written)
+}
+```
```

Comment:
I find this explanation confusing, a couple of points that come to mind:

1. You say "One purpose is to facilitate buffer re-use: it’s okay to keep around the same `WriteBuf` and repeatedly write into it.", but there's no mention of calling `WriteBuf::clear()`, or why this is preferable to a `Vec<u8>`, which you can also call `clear()` on. (As I understand it, `WriteBuf::clear()` is preferable because it always keeps some non-trivial piece of initialized memory around, while also making sure that the internal `Vec<u8>` never becomes a gigantic memory hog over the course of the program. Vs using `Vec<u8>`, for which if you call `clear()` when it's 1GB long, will remain 1GB until it's destroyed).
2. "This way, it’s easy and safe to retrieve a
mutable slice of the unfilled (but already initialized) region. This slice can be used to build
a `WriteCursor`. Another purpose of the filled region tracking is to enable multiple items to be
written consecutively in the same buffer." This assumes we want `WriteCursor` around, but it's not clear why that's a constraint.

Generally I might reframe this by starting with our constraints, then a quick overview of the `Vec<u8>` model, and then a comparison with the `WriteBuf` model, explaining how `Vec<u8>` doesn't meet those constraints but `WriteBuf` does. 

All of that said, maybe it's not worth the time clarifying right now.

Reply:
You are very right, this needs to be fleshed out further. To be honest, it’s still a WIP document, but I wanted to start writing it now since we already had some discussions in PR comments that I wanted to keep around before I forget. I’ll use your comment as reference when I get time for that :+1: